### PR TITLE
CLI command for create  and update schema

### PR DESCRIPTION
**Pull request by eloaverona**
This PR implements the CLI commands for creating and updating a schema. 
~It is based on PR #45. Commits `Add inputs/outputs to add_transaction method` onward are up for review in this PR~. 

## To test
To properly send create and update schema  transactions first you have to create a Pike Organization and Agent.

### To create an organization  

`cargo run -- -vv --url http://127.0.0.1:8080 --wait 30 -k {path_to_public_key} organization create org_id_1 my_org_name my_org_address`

### To create an agent  

`cargo run -- -vv --url http://127.0.0.1:8080 --wait 30 -k {path_to_key} agent create org_id_1  {path_to_public_key} true`

### To create a schema

`cargo run -- -vv --url http://127.0.0.1:8080  --wait 30 -k  {path_to_key} schema create {path_to_yaml_file_with_schema_definition}`

#### Example yaml for schema definition for creating new schemas:
```yaml
- name: "Lightbulb"
  description: "Example Lightbulb schema"
  properties:
    - name: "size"
      data_type: NUMBER
      description: "Lightbulb radius, in millimeters"
      number_exponent: 0
      required: true
    - name: "bulb_type"
      data_type: ENUM
      description: "Bulb type"
      enum_options: ["filament", "CF", "LED"]
      required: true
    - name: "energy_rating"
      data_type: NUMBER
      description: "EnergyStar energy rating"
      number_exponent: 0
    - name: "color"
      data_type: STRUCT
      description: "A named RGB Color value"
      struct_properties:
            - name: 'name'
              data_type: STRING
              description: "Color name"
            - name: 'rgb_hex'
              data_type: STRING
              description: "RGB value" 
- name: "Phone"
  description: "Example phone schema"
  properties:
      - name: "brand"
        data_type: STRING
        description: "Name of the brand"
        required: true
      - name: "operation_system"
        data_type: ENUM
        description: "Phone's operation system"
        enum_options: ["Android", "iOS"]
        required: true
      - name: "price"
        data_type: NUMBER
        description: "Price of phone rounded to the nearest dollar"
        number_exponent: 0
```

### To update a schema

`cargo run -- -vv --url http://127.0.0.1:8080  --wait 30 -k  {path_to_key} schema update {path_to_yaml_file_with_schema_definition}`

#### Example yaml for schema definition for updating existing schemas:
```yaml
- name: "Phone"
  properties:
    - name: "release_year"
      data_type: String
      description: "Year phone was released"
      required: true   
```

### CODE REVIEW

File: cli/src/main.rs [hunk 1]
```diff
@@ -32,7 +42,64 @@ fn run() -> Result<(), CliError> {
...
+            (about: "Update or create agent")
+            (@setting SubcommandRequiredElseHelp)
+            (@subcommand create =>
+                (about: "Create an agent")
+                (@arg org_id: +takes_value +required "organization id")
+                (@arg public_key: +takes_value +required "public key")
+                (@arg active: "Is user active")
+                (@arg roles: --roles +takes_value +multiple "Roles assigned to agent")
+                (@arg metadata: --metadata +takes_value +multiple
+                    "Comma seperated key value pairs stored in metadata")
```
Comment (chenette) [inline cli/src/main.rs]: Typo and missing hyphen: "Comma-separated".

File: cli/src/main.rs [hunk 2]
```diff
@@ -32,7 +42,64 @@ fn run() -> Result<(), CliError> {
...
+            )
+            (@subcommand update =>
+                (about: "Update an agent")
+                (@arg org_id: +takes_value +required "organization id")
+                (@arg public_key: +takes_value +required "public key")
+                (@arg active: "Is user active")
+                (@arg roles: --roles +takes_value +multiple "Roles assigned to agent")
+                (@arg metadata: --metadata +takes_value +multiple
+                    "Comma seperated key value pairs stored in metadata")
```
Comment (chenette) [inline cli/src/main.rs]: Comma-separated

File: cli/src/main.rs [hunk 3]
```diff
@@ -32,7 +42,64 @@ fn run() -> Result<(), CliError> {
...
+        (@subcommand organization =>
+            (about: "Update or create organization")
+            (@setting SubcommandRequiredElseHelp)
+            (@subcommand create =>
+                (about: "Create an organization")
+                (@arg org_id: +required +takes_value "Unique ID for organization")
+                (@arg name: +required +takes_value "Name of the organization")
+                (@arg address: +takes_value "Physical address for organization")
+                (@arg metadata: --metadata +takes_value +multiple
+                    "Comma seperated key value pairs stored in metadata")
```
Comment (chenette) [inline cli/src/main.rs]: Here, too.

File: cli/src/main.rs [hunk 4]
```diff
@@ -32,7 +42,64 @@ fn run() -> Result<(), CliError> {
...
+            )
+            (@subcommand update =>
+                (about: "Update an organization")
+                (@arg org_id: +required +takes_value "Unique ID for organization")
+                (@arg name: +required +takes_value "Name of the organization")
+                (@arg address: +takes_value "Physical address for organization")
+                (@arg metadata: --metadata +takes_value +multiple
+                    "Comma seperated key value pairs stored in metadata")
```
Comment (chenette) [inline cli/src/main.rs]: And one more.

File: cli/src/main.rs [hunk 5]
```diff
@@ -32,7 +42,64 @@ fn run() -> Result<(), CliError> {
...
+        (@arg url: --url  +takes_value "Rest api url")
+        (@arg wait: --wait +takes_value "How long to wait for transaction to be committed")
+        (@arg key: -k +takes_value "base name for private key file")
         (@arg verbose: -v +multiple "Log verbosely")
+        (@subcommand agent =>
+            (about: "Update or create agent")
+            (@setting SubcommandRequiredElseHelp)
+            (@subcommand create =>
+                (about: "Create an agent")
+                (@arg org_id: +takes_value +required "organization id")
```
Comment (chenette) [inline cli/src/main.rs]: (nitpick) id -> ID , because all caps is standard; plus, other lines use "ID"

File: cli/src/main.rs [hunk 6]
```diff
@@ -32,7 +42,64 @@ fn run() -> Result<(), CliError> {
...
+                (@arg public_key: +takes_value +required "public key")
+                (@arg active: "Is user active")
+                (@arg roles: --roles +takes_value +multiple "Roles assigned to agent")
+                (@arg metadata: --metadata +takes_value +multiple
+                    "Comma seperated key value pairs stored in metadata")
+            )
+            (@subcommand update =>
+                (about: "Update an agent")
+                (@arg org_id: +takes_value +required "organization id")
```
Comment (chenette) [inline cli/src/main.rs]: id -> ID

File: cli/src/main.rs [hunk 7]
```diff
@@ -32,7 +42,64 @@ fn run() -> Result<(), CliError> {
         (version: VERSION)
         (author: "Contributors to Hyperledger Grid")
         (about: "Command line for Hyperledger Grid")
+        (@arg url: --url  +takes_value "Rest api url")
```
Comment (chenette) [inline cli/src/main.rs]: REST API URL (all caps)

However, I recommend rewording as "URL for the REST API".

File: cli/src/main.rs [hunk 8]
```diff
@@ -32,7 +42,64 @@ fn run() -> Result<(), CliError> {
...
+        )
+        (@subcommand schema =>
+            (about: "Update or create schemas")
+            (@setting SubcommandRequiredElseHelp)
+            (@subcommand create =>
+                (about: "Create schemas from an yaml file")
+                (@arg path: +takes_value +required "Path to yaml file containing a list of schema definitions")
+            )
+            (@subcommand update =>
+                (about: "Update an schema from an yaml file")
```
Comment (chenette) [inline cli/src/main.rs]: "an schema" -> "a schema"
... and ...
"an yaml file" -> "a yaml file"

Comment (eloaverona) [reply cli/src/main.rs]: Addressed!

File: cli/src/transaction.rs [hunk 9]
```diff
@@ -38,6 +38,14 @@ pub const PIKE_NAMESPACE: &str = "cad11d";
 const PIKE_FAMILY_NAME: &str = "pike";
 const PIKE_FAMILY_VERSION: &str = "0.1";
 
+pub const GRID_NAMESPACE: &str = "621dee";
```
Comment (agunde406) [inline cli/src/transaction.rs]: Can we use the Grid schema namespace instead of just the grid namespace? 621dee01

Comment (eloaverona) [reply cli/src/transaction.rs]: Addressed!

File: cli/src/actions/schemas.rs [hunk 10]
```diff
@@ -12,13 +12,59 @@
...
+    key: Option<String>,
+    wait: u64,
+    path: &str,
+) -> Result<(), CliError> {
+    let payloads = parse_yaml(path, Action::SchemaCreate)?;
+    let mut batch_list_builder = schema_batch_builder(key);
+    for payload in payloads {
+        batch_list_builder = batch_list_builder.add_transaction(
+            &payload.into_proto()?,
+            &[PIKE_NAMESPACE.to_string(), GRID_SCHEMA_NAMESPACE.to_string()],
```
Comment (peterschwarz) [inline cli/src/actions/schemas.rs]: Need to run `cargo fmt` on this commit.

Comment (eloaverona) [reply cli/src/actions/schemas.rs]: done!

**Pull request merged**